### PR TITLE
Add Spring Boot instrumentation lab

### PR DESCRIPTION
**Pull request by phil-pona**

### CODE REVIEW

**Comment by phil-pona** [comment]
@mabegglen what do you think, should we also configure prometheus to scrape the new springboot application target, it would probably mean we would have to keep the spring boot app running, from there on?

**Comment by mabegglen** [comment]
@phil-pona I would just mention it that they can optionally add these metrics to their Prometheus, as they learned in the Labs before, and find a way to keep the app running. But not describe the whole thing in detail again.

File: content/en/docs/05/_index.md [hunk 1]
```diff
@@ -12,10 +12,135 @@ The Prometheus project provides [client libraries](https://prometheus.io/docs/in
 
 Even if you don't plan to provide your own metrics, those libraries already export some basic metrics based on the language. For [Java](https://github.com/prometheus/client_java#included-collectors), default metrics about memory management (heap, garbage collection) and thread pools can be collected. The same applies to [Go](https://prometheus.io/docs/guides/go-application/).
 
-{{% alert title="Note" color="primary" %}}
 
-Just a short mention of the Spring Framework as it is very popular in application development. The framework also supports [exporting metrics](https://spring.io/blog/2018/03/16/micrometer-spring-boot-2-s-new-application-metrics-collector) in the Prometheus data format.
-{{% /alert %}}
+### Spring Boot Example Instrumentation
+
+Using the [micrometer metrics facade](https://spring.io/blog/2018/03/16/micrometer-spring-boot-2-s-new-application-metrics-collector) in Spring Boot Applications let's us collect all sort of metrics within a Spring Boot application. Those metrics can be exported for prometheus to scrape by a few additional dependencies and configuration.
```
Comment (mabegglen) [inline content/en/docs/05/_index.md:18]:
```suggestion
Using the [micrometer metrics facade](https://spring.io/blog/2018/03/16/micrometer-spring-boot-2-s-new-application-metrics-collector) in Spring Boot Applications lets us collect all sort of metrics within a Spring Boot application. Those metrics can be exported for Prometheus to scrape by a few additional dependencies and configuration.
```

**Comment by phil-pona** [comment]
that's a good point, I'll add an additional Task for that

File: content/en/docs/05/_index.md [hunk 2]
```diff
@@ -12,10 +12,135 @@ The Prometheus project provides [client libraries](https://prometheus.io/docs/in
...
+```bash
+git clone https://github.com/acend/prometheus-training-spring-boot-example.git
+```
+
+Change into the freshly cloned git repository
+```bash
+cd prometheus-training-spring-boot-example
+```
+
+To make the application collect metrics and provide a prometheus endpoint we now need to simply add the following two dependencies in the `pom.xml` file, where it says `<!-- Add Dependencies here-->`:
```
Comment (mabegglen) [inline content/en/docs/05/_index.md:39]:
```suggestion
To make the application collect metrics and provide a Prometheus endpoint we now need to simply add the following two dependencies in the `pom.xml` file, where it says `<!-- Add Dependencies here-->`:
```

File: content/en/docs/05/_index.md [hunk 3]
```diff
@@ -12,10 +12,135 @@ The Prometheus project provides [client libraries](https://prometheus.io/docs/in
...
+            <groupId>org.springframework.boot</groupId>
+            <artifactId>spring-boot-starter-actuator</artifactId>
+        </dependency>
+        <dependency>
+            <groupId>io.micrometer</groupId>
+            <artifactId>micrometer-registry-prometheus</artifactId>
+            <scope>runtime</scope>
+        </dependency>
+```
+your `pom.xml` should look like this now.
```
Comment (mabegglen) [inline content/en/docs/05/_index.md:52]:
```suggestion
Your `pom.xml` should look like this now.
```

File: content/en/docs/05/_index.md [hunk 4]
```diff
@@ -12,10 +12,135 @@ The Prometheus project provides [client libraries](https://prometheus.io/docs/in
...
+java -jar target/prometheus-training-spring-boot-example-0.0.1-SNAPSHOT.jar
+```
+
+Verify the metrics endpoint in a different terminal:
+
+```bash
+curl http://localhost:8080/actuator/prometheus
+```
+
+expected result should look similar to
```
Comment (mabegglen) [inline content/en/docs/05/_index.md:132]:
```suggestion
Expected result should look similar to
```

File: content/en/docs/05/labs/51.md [hunk 5]
```diff
@@ -30,6 +30,17 @@ What kind of risk do you have, when you see such a metric
...
 
+### Task 3 (advanced)
+
+* Add an application metric to the Spring Boot application
+* Use a counter metric to provide the amount of requests an endpoint has been processed.
+
+
+{{% alert title="Note" color="primary" %}}
+
+This counter is just a simple example for the sake of this lab. Those kind of metrics are provided by the micrometer prometheus spring boot integration out of the box.
```
Comment (mabegglen) [inline content/en/docs/05/labs/51.md:45]:
```suggestion
This counter is just a simple example for the sake of this lab. Those kind of metrics are provided by the micrometer Prometheus Spring Boot integration out of the box.
```

File: content/en/docs/05/labs/51.md [hunk 6]
```diff
@@ -64,3 +75,83 @@ http_requests_total{status="404"} 15
...
+java -jar target/prometheus-training-spring-boot-example-0.0.1-SNAPSHOT.jar
+```
+
+Let's create a couple of requests to our new endpoint, make sure to run those commands from a second terminal window, while the Spring Boot application is still running.
+
+```bash
+curl http://localhost:8080/api
+```
+
+Then verify the prometheus metrics endpoint and look for a metric with the name `my_prometheus_instrumentation_counter_total`
```
Comment (mabegglen) [inline content/en/docs/05/labs/51.md:157]:
```suggestion
Then verify the Prometheus metrics endpoint and look for a metric with the name `my_prometheus_instrumentation_counter_total`
```

File: content/en/docs/05/labs/51.md [hunk 7]
```diff
@@ -64,3 +75,83 @@ http_requests_total{status="404"} 15
...
+
+```bash
+curl http://localhost:8080/actuator/prometheus
+```
+
+expected result:
```
Comment (mabegglen) [inline content/en/docs/05/labs/51.md:163]:
```suggestion
Expected result:
```

File: content/en/docs/05/labs/51.md [hunk 8]
```diff
@@ -64,3 +79,99 @@ http_requests_total{status="404"} 15
...
+
+We register our custom counter `myCounter` on the `MeterRegistry` in the constructor of the RestController.
+
+Then we simply increase the counter evertime the endpoint `/api` is hit. (just an example endpoint)
+
+Again build and
+```bash
+./mvnw clean package
+```
+start the Spring Boot application:
```
Comment (mabegglen) [inline content/en/docs/05/labs/51.md:145]:
```suggestion
Start the Spring Boot application:
```

**Review by mabegglen — APPROVED** [review]

**Review by mabegglen — APPROVED** [review]

**Pull request merged**